client/question/services/FeedbackGeneratorService.js

```diff
@@ -257,40 +322,87 @@ tie.factory('FeedbackGeneratorService', [
     };
 
     /**
+     * Concats two strings to create testCaseKey
+     *
+     * @param {string} testSuiteId
+     * @param {string} testCaseIndex
+     */
+    var _getTestCaseKey = function(testSuiteId, testCaseIndex) {
+      return testSuiteId + '-' + testCaseIndex;
+    };
+
+    /**
      * Returns the Feedback object created as a result of a specified
      * correctness test.
      *
-     * @param {string|null} outputFunctionName If needed, is string of the
-     *    name for the function needed to process/format the user's output.
      * @param {CorrectnessTest} TestCase
+     * @param {number} testCaseIndex
      * @param {*} observedOutput Actual output for running user's code.
+     * @param {string} testSuiteId
      * @returns {Feedback}
      * @private
      */
     var _getCorrectnessTestFeedback = function(
-      outputFunctionName, testCase, observedOutput) {
+      testCase, testCaseIndex, observedOutput, testSuiteId) {
+      var testCaseKey = _getTestCaseKey(testSuiteId, testCaseIndex);
       var allowedOutputExample = testCase.getAnyAllowedOutput();
       var feedback = FeedbackObjectFactory.create(
         FEEDBACK_CATEGORIES.INCORRECT_OUTPUT_FAILURE);
-      // TODO(eyurko): Add varied statements for when code is incorrect.
-      feedback.appendTextParagraph('Your code produced the following result:');
-      if (outputFunctionName) {
+      // Check if new / next test suite
+      if (testSuiteId !== previousTestSuiteId) {
+        previousTestSuiteId = testSuiteId;
+        // Catch regressions
+        if (correctnessTestStates.hasOwnProperty(testCaseKey)) {
+          feedback.appendTextParagraph(
+            'It looks like there was a regression in your code. Your code ' +
+            'worked for the following, but it now fails:');
+          feedback.appendCodeParagraph(
+            'Input: ' + _jsToHumanReadable(testCase.getInput()) + '\n' +
+            'Expected Output: ' + _jsToHumanReadable(allowedOutputExample));
+          return feedback;
+        }
+        correctnessTestStates[testCaseKey] = CORRECTNESS_STATE_STARTING;
+      }
+      // If the suite ID corresponds to the sample input, the question will
+      // already have displayed the input and expected output, so we advance
+      // the correctness state to "expected output displayed".
+      if (testSuiteId === TEST_SUITE_ID_SAMPLE_INPUT) {
+        correctnessTestStates[testCaseKey] =
+          CORRECTNESS_STATE_EXPECTED_OUTPUT_DISPLAYED;
+      }
+      switch (correctnessTestStates[testCaseKey]) {
+      case CORRECTNESS_STATE_STARTING:
+        // Display input user should try
```

Comment:
Display an input that the learner should use to manually walk through their code.

Also, add periods at ends of comments, here and below.

Reply:
Changed as requested.